Hazelcast.Net/Hazelcast.Client/Statistics.cs

```diff
@@ -0,0 +1,296 @@
+
+            _client.GetClientExecutionService().ScheduleWithFixedDelay(PeriodicStatisticsSendTask, _period, _period,
+                TimeUnit.Seconds, _heartbeatToken.Token);
+        }
+
+        public void Destroy()
+        {
+            if (!_live.CompareAndSet(true, false))
+            {
+                return;
+            }
+
+            _heartbeatToken.Cancel();
+        }
+
+        private ClientConnection GetOwnerConnection()
+        {
+            var ownerAddress = _client.GetClientClusterService().GetOwnerConnectionAddress();
+            var clientConnection = _client.GetConnectionManager().GetConnection(ownerAddress);
+            if (clientConnection == null)
+            {
+                return null;
+            }
+
+            if (clientConnection.ConnectedServerVersionInt >= SinceVersion)
+            {
+                return clientConnection;
+            }
+
+            // do not print too many logs if connected to an old version server
+            if (_ownerAddress == null || !_ownerAddress.Equals(ownerAddress))
+            {
+                if (Logger.IsFinestEnabled())
+                {
+                    Logger.Finest(string.Format(
+                        "Client statistics can not be sent to server {0} since, " +
+                        "connected owner server version is less than the minimum supported server version {1}",
+                        ownerAddress, SinceVersionString));
+                }
+            }
+
+            // cache the last connected server address for decreasing the log prints
+            _ownerAddress = ownerAddress;
+            return null;
+        }
+
+        private void PeriodicStatisticsSendTask()
+        {
+            var ownerConnection = GetOwnerConnection();
+            if (null == ownerConnection)
+            {
+                Logger.Finest("Can not send client statistics to the server. No owner connection.");
+                return;
+            }
+
+            var stats = new StringBuilder();
+            try
+            {
+                FillMetrics(stats, ownerConnection);
+                AddNearCacheStats(stats);
+            }
+            catch (Exception e)
+            {
+                Logger.Finest("Can not collect client statistics.", e);
+            }
+            SendStatsToOwner(stats);
+        }
+
+        private void SendStatsToOwner(StringBuilder stats)
+        {
+            var request = ClientStatisticsCodec.EncodeRequest(stats.ToString());
+            try
+            {
+                _client.GetInvocationService().InvokeOnTarget(request, _ownerAddress);
+            }
+            catch (Exception e)
+            {
+                // suppress exception, do not print too many messages
+                if (Logger.IsFinestEnabled())
+                {
+                    Logger.Finest("Could not send stats ", e);
+                }
+            }
+        }
+
+        private void RegisterMetrics()
+        {
```

Comment:
Do you think that we should test for actual runtime functioning of the gauges and not register them if the method is throwing an exception?

Reply:
We have PR builders for linux and windows so that we can test on various OS. I think we should keep it.

Reply:
What I meant by word "test" was this:
```
try {
   tryGetStat();
  registerSinceNoException();
} catch () {
 // do not register
}
```
And do the above code at production code, that is not the test code.

Reply:
Added controlled registration step